splitio/push/manager.py

```diff
@@ -247,3 +289,248 @@ def _handle_connection_end(self):
+
+    async def _event_handler(self, event):
+        """
+        Process an incoming event.
+
+        :param event: Incoming event
+        :type event: splitio.push.sse.SSEEvent
+        """
+        try:
+            parsed = self._get_parsed_event(event)
+            handle = self._event_handlers[parsed.event_type]
+        except (KeyError, EventParsingException):
+            _LOGGER.error('Parsing exception or no handler for message of type %s', parsed.event_type)
+            _LOGGER.debug(str(event), exc_info=True)
+            return
+
+        try:
+            await handle(parsed)
+        except Exception:  # pylint:disable=broad-except
+            _LOGGER.error('something went wrong when processing message of type %s',
+                          parsed.event_type)
+            _LOGGER.debug(str(parsed), exc_info=True)
+
+    async def _token_refresh(self, current_token):
+        """Refresh auth token."""
+        await asyncio.sleep(self._get_time_period(current_token))
+        await self._stop_current_conn()
+        self._running_task = asyncio.get_running_loop().create_task(self._trigger_connection_flow())
+
+    async def _get_auth_token(self):
+        """Get new auth token"""
+        try:
+            token = await self._auth_api.authenticate()
+            await self._telemetry_runtime_producer.record_token_refreshes()
+            await self._telemetry_runtime_producer.record_streaming_event((StreamingEventTypes.TOKEN_REFRESH, 1000 * token.exp,  get_current_epoch_time_ms()))
+
+        except APIException:
+            _LOGGER.error('error performing sse auth request.')
+            _LOGGER.debug('stack trace: ', exc_info=True)
+            await self._feedback_loop.put(Status.PUSH_RETRYABLE_ERROR)
+            raise
+
+        if not token.push_enabled:
+            await self._feedback_loop.put(Status.PUSH_NONRETRYABLE_ERROR)
+            raise Exception("Push is not enabled")
+
+        _LOGGER.debug("auth token fetched. connecting to streaming.")
+        return token
+
+    async def _trigger_connection_flow(self):
+        """Authenticate and start a connection."""
+        self._status_tracker.reset()
+        self._running = True
+
+        try:
+            token = await self._get_auth_token()
+            events_source = self._sse_client.start(token)
+            first_event = await anext(events_source)
+            if first_event.event == SSE_EVENT_ERROR:
+                raise(Exception("could not start SSE session"))
+
+            _LOGGER.debug("connected to streaming, scheduling next refresh")
+            self._token_task = asyncio.get_running_loop().create_task(self._token_refresh(token))
+            await self._handle_connection_ready()
+            await self._telemetry_runtime_producer.record_streaming_event((StreamingEventTypes.CONNECTION_ESTABLISHED, 0,  get_current_epoch_time_ms()))
+            await self._consume_events(events_source)
+        finally:
+            self._running = False
```

Comment:
- how does all this code deal with exceptions? what happens if fetching the token throws?
- looking at splitsse.py it doesn't look like SSE_EVENT_ERROR is ever forwarded (https://github.com/splitio/python-client/pull/396/files#diff-acd0cce8988a1ea172a3377fb16fca0aa5c1c19dab842fa1eaeabc07a1e5ab00R205-R207). do we need to check for that here? In case the first event is an error, shouldn't we handle it?
- we should cover that any line in the function throwing an exception is gracefully handled (mapped to a retryable OR a non-retryable error)
- we should not let exceptions escape tasks that are directly scheduled in the event loop, since we lose control over them, and all we'll get is a log on the client app, without being able to react or handle the error properly

Reply:
* The original code only capture exception for self._get_auth_token(), the logic is in that call, if it fails it will set the status to Status.PUSH_RETRYABLE_ERROR
* The threading code us setting splitsse status to SplitSSEClient._Status.ERRORED if first event is error, this is missing from async, I will add it. I will rework that logic between splitsse and manager

Reply:
in the threading push manager classes, if the first event is error, I only see the status is set to SplitSSEClient._Status.ERRORED, but nothing is checking that status, is it a retryable or non-retryable error?